# Step 3. Importing libraries and dataframes

In [1]:
#Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
#Import the dataframe 'customers'
#Create a path

path = r'/Users/alexbor/Desktop/Data Analytics/Immersion/Submissions/Instacart Basket Analysis'

#Import the df 

cust_df = pd.read_csv(os.path.join(path, '02 Data ', 'Original data', 'customers.csv'))

# Step 4. Exploring and wrangling the data

# 4.1. Exploring the data


In [3]:
#Wrangling the data
#Check the head

cust_df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [4]:
#Check the tail

cust_df.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [5]:
#Check the shape 

cust_df.shape

(206209, 10)

In [6]:
#Check the info

cust_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


**Note: I can already see here that there are problems with the column 'First Name'. I should keep it in mind for the next steps**

In [7]:
#Describe the numerical data 

cust_df.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

**Note: I can already see here that the income data values have a very wide scope. I should keep it in mind for now**

# 4.2. Wrangling data

In [8]:
#I decided to rename the 'Surnam' column to 'Last name' and the column 'n_dependance' to 'Number of dependants'

cust_df.rename(columns = {'Surnam':'Last Name'}, inplace = True)

In [9]:
cust_df.rename(columns = {'n_dependants':'Number of dependants'}, inplace = True)

In [10]:
#Checking the head
cust_df.head()

user_id First Name Last Name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of dependants fam_status  income  
0                     3    married  165665  
1                     0     single   59285  
2                     2    married   99568  
3                     0     single   42049  
4                     1    married   40374

# Step 5. Missing values, duplicates and data types

# 5.1. Missing values

In [11]:
#Check for missing values

cust_df.isnull().sum()

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
STATE                       0
Age                         0
date_joined                 0
Number of dependants        0
fam_status                  0
income                      0
dtype: int64

In [12]:
#Investigate the column in detail
cust_df['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

**There are 11259 missing first names, but I don't think it's relevant to do something about it. It also consitutes only 5% of the data**

# 5.2. Data types

In [13]:
#Checking for data types

cust_df.dtypes

user_id                  int64
First Name              object
Last Name               object
Gender                  object
STATE                   object
Age                      int64
date_joined             object
Number of dependants     int64
fam_status              object
income                   int64
dtype: object

In [14]:
#I decided to change the data type of user_id to string. 

cust_df['user_id'] = cust_df['user_id'].astype('str')

In [15]:
#Checking the result

cust_df['user_id'].dtypes

dtype('O')

# 5.3. Mixed-type data

In [16]:
#Checking for mixed-type data 

for col in cust_df.columns.tolist():
    weird = (cust_df[[col]].applymap(type) != cust_df[[col]].iloc[0].apply(type)).any(axis=1)
    if len (cust_df[weird]) > 0:
        print (col)

First Name


**I suppose the 'First Name' column has mixed-type data because there are NaNs there. I decided to change the data type to string**

In [17]:
#Change the data type of the column 'First Name'

cust_df['First Name'] = cust_df['First Name'].astype('str')

In [18]:
#Checking the result

cust_df['First Name'].dtypes

dtype('O')

In [19]:
#Checking again for mixed-type data 

for col in cust_df.columns.tolist():
    weird = (cust_df[[col]].applymap(type) != cust_df[[col]].iloc[0].apply(type)).any(axis=1)
    if len (cust_df[weird]) > 0:
        print (col)

# 5.4. Duplicates

In [20]:
#Finding duplicates 

df_dupl = cust_df[cust_df.duplicated()]

In [21]:
df_dupl.shape

(0, 10)

**There are no duplicates**

# Step 6. Merging the 'cust_df' df with the df from 4.8.

In [22]:
#Importing the df 'ords_prods_merged'

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data ', 'Prepared data', 'orders_products_merged_finall_with flags_4_8.pkl'))

In [23]:
#Checking the data type of 'user_id' to make sure I can merge this df with 'cust_df'

ords_prods_merge['user_id'].dtypes

dtype('int64')

In [24]:
#Changing the data type of 'user_id' in 'ords_prods_merge'

ords_prods_merge['user_id'] = ords_prods_merge['user_id'].astype('str')


In [25]:
ords_prods_merge['user_id'].dtypes

dtype('O')

In [26]:
#Drop the '_merge' column from 'ords_prods_merge'

ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [27]:
#Merging two dataframes 

ords_prods_cust_merge = cust_df.merge(ords_prods_merge, on = 'user_id', indicator=True)

In [28]:
#Check the new df info

ords_prods_cust_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   user_id                        object  
 1   First Name                     object  
 2   Last Name                      object  
 3   Gender                         object  
 4   STATE                          object  
 5   Age                            int64   
 6   date_joined                    object  
 7   Number of dependants           int64   
 8   fam_status                     object  
 9   income                         int64   
 10  product_id                     int64   
 11  product_name                   object  
 12  aisle_id                       int64   
 13  department_id                  int64   
 14  prices                         float64 
 15  order_id                       int64   
 16  order_number                   int64   
 17  orders_day_of_week       

# Step 8. Exporting the new df 

In [29]:
ords_prods_cust_merge.to_pickle(os.path.join(path, '02 Data ', 'Prepared data', 'ords_prods_cust_merge.pkl'))